src/test/java/uk/gov/pay/card/bin_ranges/BinRangeTest.java

```diff
@@ -0,0 +1,134 @@
+
+import static java.lang.String.format;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+import static org.junit.jupiter.api.Assertions.fail;
+
+class BinRangeTest {
+
+    private static final int CARD_RANGE_LENGTH = 18;
+    private static List<Range<Long>> worldpayData;
+    private static List<Range<Long>> testCardData;
+    private static List<Range<Long>> discoverData;
+
+    /**
+     * In the cardid CI pipeline, real bin ranges files will be written to the data-sources directory
+     */
+    @BeforeAll
+    static void beforeAll() {
+        URL worldpayDataSource = Thread.currentThread().getContextClassLoader().getResource("data-sources/worldpay-v3.csv");
+        URL testDataSource = Thread.currentThread().getContextClassLoader().getResource("data-sources/test-cards.csv");
+        URL discoverDataSource = Thread.currentThread().getContextClassLoader().getResource("data-sources/discover.csv");
+
+        worldpayData = loadData(worldpayDataSource, "01");
+        testCardData = loadData(testDataSource, "02");
+        discoverData = loadData(discoverDataSource, "02");
+    }
+    
+    @Test
+    void shouldHaveNoOverlapBetweenTestAndWorldpayBinRanges() {
+        checkForBinRangeOverlaps("Worldpay", worldpayData, "Test Card", testCardData);
+    }
+
+    @Test
+    void shouldHaveNoOverlapBetweenTestAndDiscoverBinRanges() {
+        checkForBinRangeOverlaps("Discover", discoverData, "Test Card", testCardData);
+    }
+
+    @Test
+    void shouldHaveNoOverlapBetweenDiscoverAndWorldpayBinRanges() {
+        checkForBinRangeOverlaps("Worldpay", worldpayData, "Discover", discoverData);
+    }
+
+    private void checkForBinRangeOverlaps(String ranges1Type, List<Range<Long>> ranges1, String ranges2Type, List<Range<Long>> ranges2) {
+        RangeSet<Long> rangeSet = TreeRangeSet.create();
+        ranges1.forEach(rangeSet::add);
+
+        List<String> overlappingRanges = new ArrayList<>();
+        ranges2.forEach(cardRange -> {
+            if (rangeSet.intersects(cardRange)) {
+                overlappingRanges.add(String.format("%s range %s intersects %s ranges %s", 
+                        ranges1Type, cardRange, ranges2Type, rangeSet.subRangeSet(cardRange)));
+            }
+        });
+
+        if (!overlappingRanges.isEmpty()) {
+            fail("Found overlaps in ranges:\n" + String.join("\n" , overlappingRanges));
+        }
+    }
+
+    /**
+      We need to check there is no overlap between ranges across all data sources because some ranges are associated
+      with corporate card numbers, as identified by the 4th "CP" column in the example line entry from the Worldpay data source:
+
+      01,511226661120000000,511226661121000000,CP,DINERS DISCOVER,.... etc
+
+      Having no overlap ranges is essential to cardid figuring out with confidence if a card is a corporate card or not.
+     */
+    @Test
+    void verifyNoOverlappingRanges() {
```

Comment:
I think this test doesn't do anything useful? It builds up the RangeSet and the HashMap from the ranges, and checks that every range in the RangeSet is in the HashMap, but that will always be the case, regardless of what the ranges are, because the ranges are added to both at the same time.

Reply:
If you wanted to test that none of the ranges overlap, you could do something like
```
@Test
void verifyNoOverlappingRanges() {
    RangeSet<Long> setOfBinRanges = TreeRangeSet.create();
    List<Range<Long>> overlappingRanges = new ArrayList<>();
    
    worldpayData.forEach(cardRange -> {
        boolean contains = setOfBinRanges.intersects(cardRange);
        if (contains) {
            overlappingRanges.add(cardRange);
        }
        
        setOfBinRanges.add(cardRange);
    });
    discoverData.forEach(cardRange -> {
        boolean contains = setOfBinRanges.intersects(cardRange);
        if (contains) {
            overlappingRanges.add(cardRange);
        }
        
        setOfBinRanges.add(cardRange);
    });
    testCardData.forEach(cardRange -> {
        boolean contains = setOfBinRanges.intersects(cardRange);
        if (contains) {
            overlappingRanges.add(cardRange);
        }
        
        setOfBinRanges.add(cardRange);
    });
    assertThat(overlappingRanges.size(), is(0));
}
```
although this somewhat mirrors the other tests - it does have the advantage of checking data within a range, not just between ranges

Reply:
Digging into the history it seems this test used to be called `checkIfNoExistingRangeInHashMap` - I'm still not clear on what the point of that test is though?

Reply:
> I think this test doesn't do anything useful? It builds up the RangeSet and the HashMap from the ranges, and checks that every range in the RangeSet is in the HashMap, but that will always be the case, regardless of what the ranges are, because the ranges are added to both at the same time.

If there are overlapping ranges, RangeSet will adjust the sets of ranges accordingly, and it will mean there will be a difference in the range sets between the RangeSet and HashMap. See how it works at https://guava.dev/releases/23.0/api/docs/com/google/common/collect/RangeSet.html. So I think your suggested code, although arguably clearer in intention, and what is currently here are essentially the same thing.

Reply:
> Digging into the history it seems this test used to be called `checkIfNoExistingRangeInHashMap` - I'm still not clear on what the point of that test is though?

Nat and I weren't clear on what the was `checkIfNoExistingRangeInHashMap` trying to achieve, so we renamed it `verifyNoOverlappingRanges` and gave it some javadoc to clarify. See https://github.com/alphagov/pay-cardid-data/pull/68/files#diff-b5d0a0f2d9eca8d9c1aea690c8dce8ab047b489f2d53c34b28987ec6f5bc31a3R56-R65